### PR TITLE
Relayer TLA+: Support for running Apalache

### PR DESCRIPTION
**Pull request by istoilkovska**
Closes:  #165 

## Description

Added support for running [Apalache](https://github.com/informalsystems/apalache/) on the relayer spec. Currently, we can check the invariant `ICS18Inv` up to 9 steps with Apalache, for height 2 and two concurrent relayers. 
Most of the changes are done in `RelayerDefinitions.tla`, where the appropriate types and operators for type annotattions are defined. These are then used in various places throught the spec. 

Note: the branch is called ilina/packets, but there isn't any support for packet handling yet.

______

For contributor use:

- [ ] Unit tests written
- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Updated relevant documentation (`docs/`) and code comments
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Review by konnov — COMMENTED** [review]
Wow. So you did not have to comment out code in the end? How long does it take apalache to check 10 steps?

File: docs/spec/relayer/RelayerDefinitions.tla [hunk 1]
```diff
@@ -90,111 +297,178 @@ ConnectionEnds ==
...
+        srcChannelID : ChannelIDs,
+        dstChannelID : ChannelIDs
+    ] <: {PacketType}
 
 (******************************** Datagrams ********************************
  A set of datagrams.
  ***************************************************************************)
-Datagrams(Heights) ==
-    [type : {"CreateClient"}, clientID : ClientIDs, height : Heights]
+Datagrams(maxHeight, maxPacketSeq) ==
```
Comment (adizere) [inline docs/spec/relayer/RelayerDefinitions.tla:322]: This is impressive! But I also interpret it as a clear sign we should not grow this spec any more.

**Review by adizere — COMMENTED** [review]
The spec looks amazing, but the complexity is above my level, so it's difficult to review.

Maybe it would be a good idea to have a call where Ilina offers a walk-through some parts of the design here, because I don't understand it enough to leave meaningful PR comments.

**Comment by istoilkovska** [comment]
> Wow. So you did not have to comment out code in the end? How long does it take apalache to check 10 steps?

Apalache gets killed at 10 steps. I'm able to check 9 steps on my machine, in 49 min. TLC checks the spec with the same constant values, up to diameter 38, in 3 min :neutral_face:

File: docs/spec/relayer/PacketHandlers.tla [hunk 2]
```diff
@@ -0,0 +1,67 @@
+--------------------------- MODULE PacketHandlers ---------------------------
```
[the remark this answers is not in the document]
Comment (istoilkovska) [reply docs/spec/relayer/PacketHandlers.tla:1]: Yeah, makes sense. I'll remove it for now, and all the dependencies to it.

File: docs/spec/relayer/RelayerDefinitions.tla [hunk 3]
```diff
@@ -5,9 +5,179 @@
...
+        packet |-> PacketType,
+        acknowledgement |-> BOOLEAN
+    ]
+
+AsID(ID) == ID <: STRING
+AsInt(n) == n <: Int
+AsString(s) == s <: STRING
+AsChannelEnd(channelEnd) == channelEnd <: ChannelEndType
+AsSetChannelEnd(CE) == CE <: {ChannelEndType}
+AsConnectionEnd(connectionEnd) == connectionEnd <: ConnectionEndType
```
[the remark this answers is not in the document]
Comment (istoilkovska) [reply docs/spec/relayer/RelayerDefinitions.tla:158]: `<:` is a type annotation operator (see [Apalache type annotations](https://github.com/informalsystems/apalache/blob/unstable/docs/types-and-annotations.md))

**Comment by konnov** [comment]
Did you try to increase the parameters and see, whether it degrades?

How about proving an inductive invariant with Apalache. Do you think it would be an insane task?

**Comment by milosevic** [comment]
If we can model check it with TLC, I think that's fine. We should wrap this up and work on packets.

**Review by milosevic — APPROVED** [review]

**Pull request merged**